speculator/train_speculator_utils.py

```diff
@@ -0,0 +1,546 @@
+import os
+import re
+import time
+from typing import Any, Callable, Mapping, MutableMapping, Optional, Tuple, Union
+
+import torch
+import torch.distributed as dist
+import torch.nn as nn
+import torch.nn.functional as F
+from fms.models import register_model
+from fms.models.gpt_bigcode import GPTBigCode
+from fms.models.gpt_bigcode import _20b_config as _gpt_bigcode_20b_config
+from fms.models.gpt_bigcode import _hf_sd_to_fms_sd as _gptbigcode_hf_sd_to_fms_sd
+from fms.models.llama import LLaMA
+from fms.models.llama import _hf_sd_to_fms_sd as _llama_hf_sd_to_fms_sd
+from fms.models.mixtral import Mixtral, MixtralConfig
+from fms.models.mixtral import _hf_sd_to_fms_sd as _mixtral_hf_sd_to_fms_sd
+from fms.utils import serialization, tokenizers
+from fms.utils.generation import _make_cache_contiguous
+from torch.nn import CrossEntropyLoss
+from torch.utils.data import DataLoader
+
+from fms_fsdp.config import train_config
+from fms_fsdp.utils.checkpointing_utils import Checkpointer
+from fms_fsdp.utils.config_utils import get_model_config
+
+
+def generate(
```

Comment:
In `fms` we added a post processing hook to generate, we may be able to re-use that and remove this method entirely. https://github.com/foundation-model-stack/foundation-model-stack/pull/324

Reply:
Ah, nvm this would be good to figure out, but as it stands currently we either return logits and past_key_values, there is no hook to get the embeds and include them.